pkg/matchfunction/matchfunction.go

```diff
@@ -0,0 +1,64 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package golang provides the Match Making Function service for Open Match golang harness.
+package matchfunction
+
+import (
+	"context"
+	"io"
+
+	"github.com/sirupsen/logrus"
+	"open-match.dev/open-match/pkg/pb"
+)
+
+var (
+	logger = logrus.WithFields(logrus.Fields{
+		"app":       "matchfunction",
+		"component": "matchfunction.library",
+	})
+)
+
+func FetchPoolTickets(ctx context.Context, mmlClient pb.MmLogicClient, req *pb.RunRequest) (map[string][]*pb.Ticket, error) {
```

Comment:
I was expecting something more like
```go
func QueryPool(ctx context.Context, mml pb.MmLogicClient, pool *pb.Pool) ([]*pb.Ticket, error) {
	query, err := mmlClient.QueryTickets(ctx, pool)
	if err != nil {
		return nil, fmt.Errorf("Error calling mmlogic.QueryTickets: %w", err)
	}

	tickets := []*pb.Ticket{}

	for {
		resp, err := query.Recv()
		if err == io.EOF {
			return tickets, nil
		}
		if err != nil {
			return nil, fmt.Errof("Error recieving tickets from mmlogic.QueryTickets: %w", err)
		}

		tickets = append(tickets, resp.Tickets...)
	}
}

func QueryPools(ctx context.Context, mml pb.MmLogicClient, pools []*pb.Pool) (map[string][]*pb.Ticket, error) {
	ctx, cancel := context.WithCancel(ctx)

	type result struct {
		err     error
		tickets []*pb.Ticket
		name    string
	}

	results := make(chan result)

	for i := range pools {
		pool := pools[i]
		go func() {
			r := result{
				name: pool.Name,
			}
			r.tickets, r.err = FetchPoolTickets(ctx, mml, pool)
			select {
			case r <- results:
			case <-ctx.Done():
			}
		}()
	}

	poolMap := make(map[string][]*pb.Ticket)

	for i := 0; i < len(pools); i++ {
		select {
		case <-ctx.Done():
			return nil, fmt.Errorf("Context canceled while querying pools: %w", ctx.Err())
		case r := <-results:
			if r.err != nil {
				cancel()
				return nil, r.err
			}
			poolMap[r.name] = r.tickets
		}
	}

	return poolMap, nil
}
```

Where there is a single method to query one pool, and querying multiple is done concurrently.

Reply:
I was also expecting the client being made as part of this package given the mmlogic address details, but I think I see now the grpc connection options might be important.

Though this will cause single mmf to query the same mmlogic because it will use one http2 connection.

Reply:
Good idea for QueryPool - and thanks for writing out the implementation :p Updated. 

I think for now, let us keep the connection out of the library - lesser magic in the library the better. I think all the examples in the tutorials will make connections to OM services - so that handled in the library does not add much.

Reply:
And yes, I think as for the http2 connection, I still think what we currently have is a makeshift solution and switching a a mesh might be the way to go. There is an issue tracking that - so when we visit it, we should either move that way or change all the examples to use the resolver client pattern and timeout connections.